File: 53-maximum-subarray/maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    let max = nums[0];
    let current = nums[0];
    let start = 0;
    let end = 0;
    for(let i=1; i<nums.length; i++){
        if(nums[i] > current + nums[i]){
            current = nums[i];
            start = i;
        } else {
            current += nums[i]; 
        }
        if(current > max){
            max = current;
            end = i;
        }
    }
    let subArr = nums.slice(start, end + 1);
    return max;
    //Reduce Method
    // subArr.reduce((acc, cur) => {
    //         return acc + cur
    // }, 0);
};

File: 53-maximum-subarray/maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int count = 0;
        int max = Integer.MIN_VALUE;
        for(int i=0; i<nums.length; i++){
            count += nums[i];
            max = count > max ? count : max;
            count = count < 0 ? 0 : count;
        }
        return max;
    }
}